File: DD/mediareader/src/main/java/com/andruid/magic/mediareader/viewmodel/ArtistViewModel.java
package com.andruid.magic.mediareader.viewmodel;

import android.app.Application;
import android.support.v4.media.MediaBrowserCompat;

import com.andruid.magic.mediareader.data.Constants;
import com.andruid.magic.mediareader.datasourcefactory.ArtistDataSourceFactory;
import com.andruid.magic.mediareader.model.Artist;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;

public class ArtistViewModel extends AndroidViewModel {
    private Application application;

    public ArtistViewModel(@NonNull Application application) {
        super(application);
        this.application = application;
    }

    public LiveData<PagedList<Artist>> getArtists(MediaBrowserCompat mediaBrowserCompat){
        PagedList.Config config = new PagedList.Config.Builder()
                .setEnablePlaceholders(false)
                .setPageSize(Constants.PAGE_SIZE)
                .build();
        return new LivePagedListBuilder<>(new ArtistDataSourceFactory(mediaBrowserCompat,application.getApplicationContext()),config).build();
    }
}

File: DiscoDruid/app/src/main/java/com/andruid/magic/discodruid/ui/viewmodel/ArtistViewModel.kt
package com.andruid.magic.discodruid.ui.viewmodel

import android.support.v4.media.MediaBrowserCompat
import androidx.lifecycle.ViewModel
import androidx.lifecycle.asLiveData
import androidx.lifecycle.liveData
import androidx.lifecycle.viewModelScope
import androidx.paging.Pager
import androidx.paging.PagingConfig
import androidx.paging.cachedIn
import com.andruid.magic.discodruid.data.PAGE_SIZE
import com.andruid.magic.discodruid.paging.ArtistsPagingSource
import kotlinx.coroutines.cancel

class ArtistViewModel(mediaBrowserCompat: MediaBrowserCompat) : ViewModel() {
    val artistLiveData = liveData {
        val config = PagingConfig(PAGE_SIZE)
        val pager = Pager(config) {
            ArtistsPagingSource(mediaBrowserCompat)
        }

        emitSource(
            pager.flow.cachedIn(viewModelScope)
                .asLiveData()
        )
    }

    override fun onCleared() {
        super.onCleared()
        viewModelScope.cancel()
    }
}